src/commands/add.js

```diff
@@ -0,0 +1,208 @@
+      )
+      process.exitCode = 1
+      return
+    }
+
+    let ethabi = null
+    if (abi) {
+      ethabi = EthereumABI.load(contractName, abi)
+    } else {
+      if (network === 'poa-core') {
+        ethabi = await loadAbiFromBlockScout(EthereumABI, network, address)
+      } else {
+        ethabi = await loadAbiFromEtherscan(EthereumABI, network, address)
+      }
+    }
+
+    let updateResult = updateEventNamesOnCollision(ethabi, entities, contractName, mergeEntities)
+    let collisionEntities = updateResult.collisionEntities
+    let onlyCollisions = updateResult.onlyCollisions
+    ethabi.data = updateResult.abiData
+
+    await writeABI(ethabi, contractName, abi)
+    await writeSchema(ethabi, protocol, result.getIn(['schema', 'file']), collisionEntities)
+    await writeMapping(ethabi, protocol, contractName, collisionEntities)
+
+    let dataSources = result.get('dataSources')
+    let dataSource = await generateDataSource(protocol, 
+      contractName, network, address, ethabi)
+
+    // Handle the collisions edge case by copying another data source yaml data
+    if (mergeEntities && onlyCollisions) {
+      let firstDataSource = dataSources.get(0)
+      let source = dataSource.get('source')
+      let mapping = firstDataSource.get('mapping').asMutable()
+
+      // Save the address of the new data source
+      source.abi = firstDataSource.get('source').get('abi')
+
+      dataSource.set('mapping', mapping)
+      dataSource.set('source', source)
+    }
+
+    result.set('dataSources', dataSources.push(dataSource))
+
+    await Subgraph.write(result, manifestPath)
+    manifest = await Subgraph.load(manifestPath, {protocol: protocol})
+
+    // Detect Yarn and/or NPM
+    let yarn = await system.which('yarn')
+    let npm = await system.which('npm')
+    if (!yarn && !npm) {
+      print.error(
+        `Neither Yarn nor NPM were found on your system. Please install one of them.`,
+      )
+      process.exitCode = 1
+      return
+    }
+
+    await withSpinner(
+      'Running codegen',
+      'Failed to run codegen',
+      'Warning during codegen',
+      async spinner => {
+        await system.run(yarn ? 'yarn codegen' : 'npm run codegen')
+      }
+    )
+  }
+}
+
+const getEntities = (manifest) => {
+  let list = []
+  manifest.result.get('dataSources').map(dataSource => {
+    dataSource.getIn(['mapping', 'entities']).map(entity => {
+      list.push(entity)
+    })
+  })
+  manifest.result.get('templates').map(template => {
+    template.getIn(['mapping', 'entities']).map(entity => {
+      list.push(entity)
+    })
+  })
+  return list
+}
+
+const getContractNames = (manifest) => {
+  let list = []
+  manifest.result.get('dataSources').map(dataSource => {
+    list.push(dataSource.get('name'))
+  })
+  manifest.result.get('templates').map(template => {
+    list.push(template.get('name'))
+  })
+  return list
+}
```

Comment:
I'm not the biggest fan of immutable.js, but since this project is using it, I think it would be better to either change those `.map`s to `.forEach`es since you're not using the new generated list, or they could use a `concat` + `map`:

```javascript
const getEntities = (manifest) => {
  let dataSources = manifest.result.get('dataSources')
  let templates = manifest.result.get('templates')

  return dataSources
    .concat(templates)
    .map(dataSource => dataSource.getIn(['mapping', 'entities']))
    .flatten()
}

const getContractNames = (manifest) => {
  let dataSources = manifest.result.get('dataSources')
  let templates = manifest.result.get('templates')

  return dataSources
    .concat(templates)
    .map(dataSource => dataSource.get('name'))
}
```

If for any reason there's issues with empty `List`s, `.get` receives a second argument that is the default value, which could be just a `List()`.

Reply:
I added the empty list arg as the default return type is `undefined`